### PR TITLE
[CECO-1067] DDA config for Admission Controller sidecar injection using selector and profile

### PR DESCRIPTION
**Pull request by kisungyi92**
### What does this PR do?

This PR is following up PR #1207 adding selectors and profiles.

### Motivation

CECO-1067

### Additional Notes

Anything else we should know when reviewing?

### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: vX.Y.Z
* Cluster Agent: vX.Y.Z

### Describe your test plan

Test based on the release v1.6.0
- Run DDA with `selectors.objectSelector` option in admissionController.agentSidecarInjection.
- Test `selectors.namespaceSelector` with same structure.

```
  features:
        selectors:
        - objectSelector:
            matchLabels:
              "testLabel": "testValue"
```
```
  features:
        selectors:
        - namespaceSelector:
            matchLabels:
              "testLabel": "testValue"
```

- Check that the `DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_SELECTORS` is set to correct map structure.
- Run command : kubectl describe deploy datadog-cluster-agent | grep -i DD_ADMISSION_CONTROLLER_AGENT_SIDECAR

```
Result #1
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_SELECTORS:              [{"objectSelector":{"matchLabels":{"testLabel":"testValue"}}}]
Result #2
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_SELECTORS:              [{"namespaceSelector":{"matchLabels":{"testLabel":"testValue"}}}]

```

- Next, test DDA with `profiles` option in admissionController.agentSidecarInjection.

```
  features:
    admissionController:
      agentSidecarInjection:
        enabled: true
        provider: "fargate"
        profiles:
        - env:
          - name: DD_PROCESS_AGENT_PROCESS_COLLECTION_ENABLED
            value: "true"
          resources:
            requests:
              cpu: "400m"
              memory: "256Mi"
            limits:
              cpu: "800m"
              memory: "512Mi"
```

- Check that the `DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROFILES` is set to correct map structure.
- Run command : kubectl describe deploy datadog-cluster-agent | grep -i DD_ADMISSION_CONTROLLER_AGENT_SIDECAR

```
Result : 
     DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROFILES:               [{"env":[{"name":"DD_PROCESS_AGENT_PROCESS_COLLECTION_ENABLED","value":"true"}],"resources":{"limits":{"cpu":"800m","memory":"512Mi"},"requests":{"cpu":"400m","memory":"256Mi"}}}]

```

### Checklist

- [X] PR has at least one valid label: `bug`, `enhancement`, `refactoring`, `documentation`, `tooling`, and/or `dependencies`
- [X] PR has a milestone or the `qa/skip-qa` label

### CODE REVIEW

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 1]
```diff
@@ -1296,6 +1296,29 @@ type AgentSidecarInjectionConfig struct {
 	// Image overrides the default Agent image name and tag for the Agent sidecar.
 	// +optional
 	Image *commonv1.AgentImageConfig `json:"image,omitempty"`
+	//Selectors defines the pod selector for sidecar injection, currently only one rule is supported.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:1299]:
```suggestion
	// Selectors defines the pod selector for sidecar injection. Currently only one rule is supported.
	// +optional
	// +listType=atomic
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 2]
```diff
@@ -1296,6 +1296,29 @@ type AgentSidecarInjectionConfig struct {
...
+	Selectors []*Selector `json:"selectors,omitempty"`
+	//profiles defines the sidecar configuration override, currently only one profile is supported.
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:1301]:
```suggestion
	// Profiles defines the sidecar configuration override. Currently only one profile is supported.
	// +optional
	// +listType=atomic
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 3]
```diff
@@ -1296,6 +1296,29 @@ type AgentSidecarInjectionConfig struct {
...
+	Profiles []*Profile `json:"profiles,omitempty"`
+}
+
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:1304]:
```suggestion
// Selectors define a pod selector for sidecar injection.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 4]
```diff
@@ -1296,6 +1296,29 @@ type AgentSidecarInjectionConfig struct {
...
+type Selector struct {
+	// NamespaceSelector specifies the label selector for namespaces.
+	// +option
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+	// ObjectSelector specifies the label selector for objects.
+	// +optional
+	ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty"`
+}
+
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:1313]:
```suggestion
// Profile defines a sidecar configuration override.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 5]
```diff
@@ -1296,6 +1296,29 @@ type AgentSidecarInjectionConfig struct {
...
 	Image *commonv1.AgentImageConfig `json:"image,omitempty"`
+	//Selectors defines the pod selector for sidecar injection, currently only one rule is supported.
+	Selectors []*Selector `json:"selectors,omitempty"`
+	//profiles defines the sidecar configuration override, currently only one profile is supported.
+	Profiles []*Profile `json:"profiles,omitempty"`
+}
+
+type Selector struct {
+	// NamespaceSelector specifies the label selector for namespaces.
+	// +option
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:1307]:
```suggestion
	// +optional
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 6]
```diff
@@ -1296,6 +1296,29 @@ type AgentSidecarInjectionConfig struct {
...
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+	// ObjectSelector specifies the label selector for objects.
+	// +optional
+	ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty"`
+}
+
+type Profile struct {
+	// EnvVars specifies the environment variables for the profile.
+	// +optional
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:1316]:
```suggestion
	// +optional
	// +listType=map
	// +listMapKey=name
```

File: controllers/datadogagent/feature/admissioncontroller/feature.go [hunk 7]
```diff
@@ -155,6 +159,48 @@ func (f *admissionControllerFeature) Configure(dda *v2alpha1.DatadogAgent) (reqC
 			} else if ok && componentOverride.Image != nil {
 				f.agentSidecarConfig.imageTag = componentOverride.Image.Tag
 			}
+
+			// check selector slices If they have values.
```
Comment (celenechang) [inline controllers/datadogagent/feature/admissioncontroller/feature.go:163]:
```suggestion
			// Assemble agent sidecar selectors
```

File: controllers/datadogagent/feature/admissioncontroller/feature.go [hunk 8]
```diff
@@ -155,6 +159,48 @@ func (f *admissionControllerFeature) Configure(dda *v2alpha1.DatadogAgent) (reqC
...
+						newSelector.ObjectSelector = objSelector
+					}
+				}
+
+				if newSelector.NamespaceSelector != nil || newSelector.ObjectSelector != nil {
+					f.agentSidecarConfig.selectors = append(f.agentSidecarConfig.selectors, newSelector)
+				}
+			}
+
+			// check profile slices If they have values.
```
Comment (celenechang) [inline controllers/datadogagent/feature/admissioncontroller/feature.go:192]:
```suggestion
			// Assemble agent sidecar profiles
```

File: controllers/datadogagent/feature/admissioncontroller/feature.go [hunk 9]
```diff
@@ -155,6 +159,48 @@ func (f *admissionControllerFeature) Configure(dda *v2alpha1.DatadogAgent) (reqC
...
+				newSelector := &v2alpha1.Selector{}
+
+				if selector.NamespaceSelector != nil {
+					nsSelector := &metav1.LabelSelector{
+						MatchLabels:      selector.NamespaceSelector.MatchLabels,
+						MatchExpressions: selector.NamespaceSelector.MatchExpressions,
+					}
+					if len(nsSelector.MatchLabels) > 0 || len(nsSelector.MatchExpressions) > 0 {
+						newSelector.NamespaceSelector = nsSelector
+					}
```
Comment (celenechang) [inline controllers/datadogagent/feature/admissioncontroller/feature.go:174]: Could we instead consolidate this?
```suggestion
				if selector.NamespaceSelector != nil {
					newSelector.NamespaceSelector := &metav1.LabelSelector{
						MatchLabels:      selector.NamespaceSelector.MatchLabels,
						MatchExpressions: selector.NamespaceSelector.MatchExpressions,
					}
```

File: controllers/datadogagent/feature/admissioncontroller/feature.go [hunk 10]
```diff
@@ -308,6 +354,28 @@ func (f *admissionControllerFeature) ManageClusterAgent(managers feature.PodTemp
 			})
 		}
 
+		if f.agentSidecarConfig.selectors != nil {
```
Comment (celenechang) [inline controllers/datadogagent/feature/admissioncontroller/feature.go:357]: I think here you want to check for length, not `nil`
```suggestion
		if len(f.agentSidecarConfig.selectors) > 0 {
```

File: docs/configuration.v2alpha1.md [hunk 11]
```diff
@@ -41,8 +41,10 @@ spec:
 | features.admissionController.agentSidecarInjection.image.pullPolicy | The Kubernetes pull policy: Use Always, Never, or IfNotPresent. |
 | features.admissionController.agentSidecarInjection.image.pullSecrets | It is possible to specify Docker registry credentials. See https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod |
 | features.admissionController.agentSidecarInjection.image.tag | Define the image tag to use. To be used if the Name field does not correspond to a full image string. |
+| features.admissionController.agentSidecarInjection.profiles | Profiles defines the sidecar configuration override. Currently only one profile is supported. |
```
Comment (maycmlee) [inline docs/configuration.v2alpha1.md:44]:
```suggestion
| features.admissionController.agentSidecarInjection.profiles | Profiles define the sidecar configuration override. Only one profile is supported. |
```

File: docs/configuration.v2alpha1.md [hunk 12]
```diff
@@ -41,8 +41,10 @@ spec:
...
 | features.admissionController.agentSidecarInjection.provider | Provider is used to add infrastructure provider-specific configurations to the Agent sidecar. Currently only "fargate" is supported. To use the feature in other environments (including local testing) omit the config. See also: https://docs.datadoghq.com/integrations/eks_fargate |
 | features.admissionController.agentSidecarInjection.registry | Registry overrides the default registry for the sidecar Agent. |
+| features.admissionController.agentSidecarInjection.selectors | Selectors defines the pod selector for sidecar injection. Currently only one rule is supported. |
```
Comment (maycmlee) [inline docs/configuration.v2alpha1.md:47]:
```suggestion
| features.admissionController.agentSidecarInjection.selectors | Selectors define the pod selector for sidecar injection. Only one rule is supported. |
```

**Review by maycmlee — COMMENTED** [review]
A couple of small suggestions

**Review by maycmlee — APPROVED** [review]

**Review by levan-m — APPROVED** [review]

**Pull request merged**